src/Main.purs

```diff
@@ -2,33 +2,60 @@ module Main where
 
 import Prelude
 
-import Control.Monad.Aff (Fiber, launchAff)
+import Dashboard.Component as Dash
+import Control.Monad.Aff (Aff, Milliseconds(..), delay)
 import Control.Monad.Aff.Console (CONSOLE, log)
 import Control.Monad.Eff (Eff)
-import DOM (DOM)
-import Gitlab (getProjects, getJobs, BaseUrl(..), Token(..))
+import Dashboard.Model as Model
+import Data.Array (uncons)
+import Data.Maybe (Maybe(..))
+import Gitlab (BaseUrl(..), Token(..), getJobs, getProjects)
+import Global.Unsafe (unsafeStringify)
+import Halogen as H
+import Halogen.Aff as HA
+import Halogen.VDom.Driver (runUI)
 import Network.HTTP.Affjax (AJAX)
-import Simple.JSON (writeJSON)
 import URLSearchParams as URLParams
 
-type Main = forall e.
-            Eff ( ajax :: AJAX
-                , console :: CONSOLE
-                , dom :: DOM
-                | e)
-                (Fiber ( ajax :: AJAX
-                       , console :: CONSOLE
-                       , dom :: DOM
-                       | e)
-                       Unit)
 
-main :: Main
+pollProjects ::
+  forall a eff.
+  BaseUrl
+  -> Token
+  -> (Dash.Query Unit -> Aff (ajax :: AJAX, console :: CONSOLE | eff) a)
+  -> Aff (ajax :: AJAX, console :: CONSOLE | eff) Unit
+pollProjects baseUrl token query = do
+  -- Get projects, poll all of them for jobs
+  log "Fetching list of projects..."
+  projects <- getProjects baseUrl token
+  fetchJobs projects
+
+  -- Wait 30 secs and recur
+  delay (Milliseconds 30000.0)
+  _ <- pollProjects baseUrl token query
+  pure unit
+  where
+    -- If list of projects is empty, return
+    -- If not, take the first, get jobs, and upsert in the Component
+    -- Then wait 1s, and recur to fetch the rest
+    fetchJobs projects = case uncons projects of
+      Nothing -> pure unit
+      Just { head: p, tail: ps } -> do
```

Comment:
`Maybe` is `Traversable`, so instead of pattern matching here you can use this beautiful `for_` thing here like:

```purescript
for_ (uncons projects) \{head: p, tail: ps } -> do
  ...
```

Reply:
On the other hand, explicit pattern matching reads well and explicit along with the comment on top of it.

Reply:
`for_` is cool, but I also agree with having things explicit, so I vote for the pattern matching.

Reply:
👍